### PR TITLE
New Time and Date Management

### PR DESCRIPTION
**Pull request by zhuyaliang**

### CODE REVIEW

File: capplets/time-admin/data/mate-time-admin.desktop.in [hunk 1]
```diff
@@ -0,0 +1,13 @@
+[Desktop Entry]
+Name=Time And Date Manager
+Name[zh_CN]=时间和日期管理
+Comment=Manage Time And Date
```
Comment (rbuj) [inline capplets/time-admin/data/mate-time-admin.desktop.in]: Don't use Title Case for Comment in mate-time-admin.desktop.in:
```shell
$ git grep "Comment="
capplets/about-me/mate-about-me.desktop.in:_Comment=Set your personal information
capplets/accessibility/at-properties/mate-at-properties.desktop.in:_Comment=Choose which accessibility features to enable when you log in
capplets/appearance/data/mate-appearance-properties.desktop.in:_Comment=Customize the look of the desktop
capplets/appearance/data/mate-theme-installer.desktop.in:_Comment=Installs themes packages for various parts of the desktop
capplets/appearance/theme-save.c:               "Comment=%s\n"
capplets/default-applications/mate-default-applications-properties.desktop.in:_Comment=Select your default applications
capplets/display/mate-display-properties.desktop.in:_Comment=Change resolution and position of monitors and projectors
capplets/keybindings/mate-keybinding.desktop.in:_Comment=Assign shortcut keys to commands
capplets/keyboard/mate-keyboard.desktop.in:_Comment=Set your keyboard preferences
capplets/mouse/mate-settings-mouse.desktop.in:_Comment=Set your mouse preferences
capplets/network/mate-network-properties.desktop.in:_Comment=Set your network proxy preferences
capplets/time-admin/data/mate-time-admin.desktop.in:Comment=Manage Time And Date 
capplets/windows/mate-window-properties.desktop.in:_Comment=Set your window properties
font-viewer/mate-font-viewer.desktop.in:_Comment=View fonts on your system
shell/matecc.desktop.in:_Comment=Configure MATE settings
```
Name and Comment should start with underscore for l10n. Test:
```shell
$ ./autogen.sh --prefix=/usr
$ ./makepot
$ grep "Time And Date Manager" mate-control-center.pot
(changes)
$ ./makepot
$ grep "Time And Date Manager" mate-control-center.pot
```

Remove trailing space in Comment.


*.desktop.in files shouldn't include Name[LOCALE]=TRANSLATION, Comment[LOCALE]=TRANSLATION

Comment (zhuyaliang) [reply capplets/time-admin/data/mate-time-admin.desktop.in]: Has been modified

Comment (rbuj) [reply capplets/time-admin/data/mate-time-admin.desktop.in]: - Don't use Title Case for Comment: Unlike Name which is a title, Comment is a sentence, for that reason you can't use Title Case for Comment.

See other _Comment= lines:
```shell
$ git grep "_Comment="
capplets/about-me/mate-about-me.desktop.in:_Comment=Set your personal information
capplets/accessibility/at-properties/mate-at-properties.desktop.in:_Comment=Choose which accessibility features to enable when you log in
capplets/appearance/data/mate-appearance-properties.desktop.in:_Comment=Customize the look of the desktop
capplets/appearance/data/mate-theme-installer.desktop.in:_Comment=Installs themes packages for various parts of the desktop
capplets/default-applications/mate-default-applications-properties.desktop.in:_Comment=Select your default applications
capplets/display/mate-display-properties.desktop.in:_Comment=Change resolution and position of monitors and projectors
capplets/keybindings/mate-keybinding.desktop.in:_Comment=Assign shortcut keys to commands
capplets/keyboard/mate-keyboard.desktop.in:_Comment=Set your keyboard preferences
capplets/mouse/mate-settings-mouse.desktop.in:_Comment=Set your mouse preferences
capplets/network/mate-network-properties.desktop.in:_Comment=Set your network proxy preferences
capplets/time-admin/data/mate-time-admin.desktop.in:_Comment=Manage Time And Date 
capplets/windows/mate-window-properties.desktop.in:_Comment=Set your window properties
font-viewer/mate-font-viewer.desktop.in:_Comment=View fonts on your system
shell/matecc.desktop.in:_Comment=Configure MATE settings
```

- Remove trailing space in Comment.

Test:
```shell
$  ./autogen.sh --prefix=/usr && ./makepot 
$  grep -A 1 mate-time-admin mate-control-center.pot
#: ../capplets/time-admin/data/mate-time-admin.desktop.in.h:1
msgid "Time And Date Manager"
--
#: ../capplets/time-admin/data/mate-time-admin.desktop.in.h:2
msgid "Manage Time And Date "
```

Comment (zhuyaliang) [reply capplets/time-admin/data/mate-time-admin.desktop.in]: Thank you for your advice.

File: capplets/time-admin/src/time-tool.c [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+*   You should have received a copy of the GNU General Public License
+*   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include "time-tool.h"
+
+struct tm *GetCurrentTime(void)
```
Comment (rbuj) [inline capplets/time-admin/src/time-tool.c]: Remove `struct tm *GetCurrentTime(void)` and use [GDateTime *g_date_time_new_now_local (void);](https://developer.gnome.org/glib/stable/glib-GDateTime.html#g-date-time-new-now-local) and adapt the code.

File: capplets/time-admin/src/time-tool.c [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+                                  G_DBUS_CALL_FLAGS_NONE,
+                                  -1,
+                                  NULL,
+                                  &error);
+
+    if(ret == NULL)
+    {
+        MessageReport(_("Set time zone"),error->message,ERROR);
+    }    
+    else
```
Comment (rbuj) [inline capplets/time-admin/src/time-tool.c]: Remove this else

File: capplets/time-admin/src/time-share.h [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <time.h>
+#include <sys/time.h>
+#include <locale.h>
+#include <libintl.h>
+#include <gio/gio.h>
+
+#define LOCALEDIR  "/usr/share/locale/"
```
Comment (rbuj) [inline capplets/time-admin/src/time-share.h]: Setting LOCALEDIR here makes source not portable (POSIX).

File: capplets/time-admin/src/time-tool.c [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+                                 "SetTimezone",
+                                  g_variant_new ("(sb)",zone,1),
+                                  G_DBUS_CALL_FLAGS_NONE,
+                                  -1,
+                                  NULL,
+                                  &error);
+
+    if(ret == NULL)
+    {
+        MessageReport(_("Set time zone"),error->message,ERROR);
```
Comment (rbuj) [inline capplets/time-admin/src/time-tool.c]: Memory leak: Missing `g_error_free(error);`

File: capplets/time-admin/src/time-tool.c [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+                                 "SetNTP",
+                                  g_variant_new ("(bb)",state,state),
+                                  G_DBUS_CALL_FLAGS_NONE,
+                                  -1,
+                                  NULL,
+                                  &error);
+
+    if(ret1 == NULL || ret2 == NULL)
+    {
+        MessageReport(_("Reload Ntp sync"),error->message,ERROR);
```
Comment (rbuj) [inline capplets/time-admin/src/time-tool.c]: Memory leak: Missing `g_error_free(error);`

File: capplets/time-admin/src/time-tool.c (hunk 2 above)
Comment (zhuyaliang) [reply capplets/time-admin/src/time-tool.c]: Why not use ```struct tm```

File: capplets/time-admin/src/time-tool.c [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+                                 "SetNTP",
+                                  g_variant_new ("(bb)",state,state),
+                                  G_DBUS_CALL_FLAGS_NONE,
+                                  -1,
+                                  NULL,
+                                  &error);
+
+    if(ret == NULL)
+    {
+        MessageReport(_("Set Ntp sync"),error->message,ERROR);
```
Comment (rbuj) [inline capplets/time-admin/src/time-tool.c]: Memory leak: Missing `g_error_free(error);`

File: capplets/time-admin/src/time-tool.c [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+                                 "SetTime",
+                                  g_variant_new ("(xbb)",TimeSec * 1000 * 1000,0,0),
+                                  G_DBUS_CALL_FLAGS_NONE,
+                                  -1,
+                                  NULL,
+                                  &error);
+
+    if(ret == NULL)
+    {
+        MessageReport(_("Set Ntp sync"),error->message,ERROR);
```
Comment (rbuj) [inline capplets/time-admin/src/time-tool.c]: Memory leak: Missing `g_error_free(error);`

File: capplets/time-admin/src/time-tool.c [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+    }    
+    else
```
Comment (rbuj) [inline capplets/time-admin/src/time-tool.c]: Remove this else

File: capplets/time-admin/src/time-tool.c (hunk 2 above)
Comment (rbuj) [reply capplets/time-admin/src/time-tool.c]: [GDateTime](https://developer.gnome.org/glib/stable/glib-GDateTime.html) is based on object-oriented programming and it has useful functions, like g_date_time_get_year (GDateTime *datetime) that doesn't require to add a base year (1900).

Comment (zhuyaliang) [reply capplets/time-admin/src/time-tool.c]: Thanks for your help If  ```struct tm``` has no other big problems, I don't think it's necessary to replace it.

**Comment by rbuj** [comment]
All time zones display the same time difference to GMT/UTC (UTC+2), and the same time (13:10):

Captura de pantalla a 2019-06-10 13-10-50

File: capplets/time-admin/src/time-share.c [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    gtk_dialog_add_action_widget (GTK_DIALOG (dialog), button, response_id);
+
+    return button;
+}
+void SetTooltip(GtkWidget*box,gboolean mode)
+{
+    if(!mode)
+    {
+        gtk_widget_set_tooltip_markup(box,
+                                     _("Network time synchronization has been set up.prohibit Modification Time"));
```
Comment (rbuj) [inline capplets/time-admin/src/time-share.c]: Grammar mistake:
Network time synchronization has been set up. Manual time/date setting is disabled.

File: capplets/time-admin/src/time-tool.c [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+{
+	time_t tt;
+	tzset();
+	tt=time(NULL);
+	
+    return localtime(&tt);
+}
+static void UpdateDate(TimeAdmin *ta,gboolean state)
+{
+    struct tm *LocalTime;
```
Comment (rbuj) [inline capplets/time-admin/src/time-tool.c]: **cppcheck warning**
Id: variableScope
The scope of the variable 'LocalTime' can be reduced.

File: capplets/time-admin/src/main.c [hunk 12]
```diff
@@ -0,0 +1,430 @@
...
+*        
+*        
+* Output:  start        :TRUE
+*          not start    :FALSE
+*        
+* Author:  zhuyaliang  31/07/2018
+******************************************************************************/
+static gboolean ProcessRuning(void)
+{
+    int fd = 0;
```
Comment (rbuj) [inline capplets/time-admin/src/main.c]: **cppcheck warning**
Id: unreadVariable
Variable 'fd' is assigned a value that is never used.

File: capplets/time-admin/src/time-tool.c (hunk 11 above)
Comment (zhuyaliang) [reply capplets/time-admin/src/time-tool.c]: I don't think it needs to be changed. It's just a matter of programming habits.

File: capplets/time-admin/src/main.c (hunk 12 above)
Comment (zhuyaliang) [reply capplets/time-admin/src/main.c]: Can't this be initialized to zero?

**Comment by zhuyaliang** [comment]
@rbuj 
```All time zones display the same time difference to GMT/UTC (UTC+02), and the same time (13:10)```
I don't quite understand the meaning of this question.Can you elaborate on it?

File: capplets/time-admin/src/main.c (hunk 12 above)
Comment (rbuj) [reply capplets/time-admin/src/main.c]: fd initialization is overridden in line 166, and the initialization value (0) is not used:
```C
fd = open(LOCKFILE,O_RDONLY);
```
It's a cppcheck warning, we suggest to developers to keep [build, cppcheck and clang reports](https://mate-control-center.mate-desktop.dev/) as clean as possible.

**Comment by rbuj** [comment]
@zhuyaliang 
I launched mate-time-admin from a command line, then I clicked on button next to "Time Zone:" label, which has the current timezone.

Next, I clicked on two different time zones, but the time displayed as well as the time diff with GMT have always the same value. I think that Time Zone Selection dialog shows the current time of the selected timezone when the user clicks on any timezone. I think that this is the expected behavior.

**Comment by zhuyaliang** [comment]
@rbuj Okay, that's a very good suggestion. I'll make some changes

File: capplets/time-admin/src/main.c (hunk 12 above)
Comment (zhuyaliang) [reply capplets/time-admin/src/main.c]: I will try to get rid of these warnings

**Comment by zhuyaliang** [comment]
@rbuj How does it look now?

File: capplets/time-admin/src/time-zone.c [hunk 13]
```diff
@@ -269,17 +262,22 @@ static void
...
     TzLocation       *current_location;
     GDateTime        *date;
 
-    date = g_date_time_new_now_local ();
     current_location = timezone_map_get_location (TIMEZONEMAP (map));
     city_country = translated_city_name (current_location);
     
+    timezone = g_time_zone_new (current_location->zone);
+    current_date = g_date_time_new_now_local ();
+    date = g_date_time_to_timezone (current_date, timezone);
```
Comment (rbuj) [inline capplets/time-admin/src/time-zone.c]: You should release the return value of[ g_date_time_to_timezone()](https://developer.gnome.org/glib/stable/glib-GDateTime.html#g-date-time-to-timezone) by calling [g_date_time_unref()](https://developer.gnome.org/glib/stable/glib-GDateTime.html#g-date-time-unref) when you are done with it.

**Comment by rbuj** [comment]
@zhuyaliang It looks fantastic! I love it! You have done a great job.

File: capplets/time-admin/src/time-zone.c (hunk 13 above)
Comment (zhuyaliang) [reply capplets/time-admin/src/time-zone.c]: It's my oversight. I've changed the problem.

**Comment by zhuyaliang** [comment]
> @zhuyaliang看起来很棒！我喜欢它！你已经做了很多工作。

Thank you for your help, so that he will be more perfect.

**Comment by zhuyaliang** [comment]
@yetist  @raveit65 
Are there any other suggestions?

File: capplets/time-admin/src/main.c [hunk 14]
```diff
@@ -0,0 +1,430 @@
...
+*   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "time-tool.h"
+#include "time-zone.h"
+#include "time-map.h"
+
+#define  LOCKFILE               "/tmp/time-admin.pid"
+#define  TIME_ADMIN_PERMISSION  "org.mate.user.admin.administration"
+#define  ICONFILE               "/usr/share/mate-time-admin/icon/mate-times-admin.png"
```
Comment (yetist) [inline capplets/time-admin/src/main.c]: Maybe you can replace  `/usr/share` with DATADIR?

File: capplets/time-admin/src/time-map.c [hunk 15]
```diff
@@ -0,0 +1,629 @@
...
+
+    /* paint background */
+    gdk_cairo_set_source_pixbuf (cr, map->background, 0, 0);
+    cairo_paint (cr);
+
+    /* paint hilight */
+    if (gtk_widget_is_sensitive (widget))
+    {
+        file = g_strdup_printf ("/usr/share/mate-time-admin/map/timezone_%s.png",
+                                g_ascii_formatd (buf, sizeof (buf),
```
Comment (yetist) [inline capplets/time-admin/src/time-map.c]: There is a hard-coded path, maybe you can use pkgdatadir?

Comment (zhuyaliang) [reply capplets/time-admin/src/time-map.c]: Now use the definition in Makefile

File: capplets/time-admin/src/time-map.c [hunk 16]
```diff
@@ -333,13 +333,13 @@ cc_timezone_map_draw (GtkWidget *widget,
     /* paint hilight */
     if (gtk_widget_is_sensitive (widget))
     {
-        file = g_strdup_printf ("/usr/share/mate-time-admin/map/timezone_%s.png",
+        file = g_strdup_printf (TIMPZONEDIR"timezone_%s.png",
```
Comment (zhuyaliang) [inline capplets/time-admin/src/time-map.c]: Do you mean this?

**Comment by zhuyaliang** [comment]
@lukefromdc 
First, there is an NTP error. Maybe there is something wrong with your systemd version. I used the timedada1 Dbus service provided by systemd to complete the time modification and synchronization.

I can change the second color to blue. How do you feel?

In addition, the service does not require authentication

**Comment by raveit65** [comment]
I agree with @lukefromdc  that it is hard to read the words `Time Zone` and `Ntp Sync`.
For example with light (Blue)Menta which is our default theme.
Bildschirmfoto zu 2019-06-12 10-34-41
I didn't found a hint with `gtk3-inspector` why this labels doesn't use normal fg_color from theme which is black.
Also `Set Time` and `Set Date` should use normal fg_color from theme, ihmo.

**Comment by zhuyaliang** [comment]
@raveit65 
You mean that labels here use default colors and fonts. For example, the color of the text in button

**Comment by zhuyaliang** [comment]
For example
2019-06-12 17-21-43屏幕截图

**Comment by raveit65** [comment]
Yes, this is better readable.

File: capplets/time-admin/src/time-share.c [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+    gtk_message_dialog_format_secondary_markup(GTK_MESSAGE_DIALOG(dialog),
+                                               TYPEMSG,
+                                               Msg);
+    gtk_window_set_title(GTK_WINDOW(dialog),("Message"));
+    nRet =  gtk_dialog_run(GTK_DIALOG(dialog));
+    gtk_widget_destroy(dialog);
+
+    return nRet;
+}
+void SetLableFontType(GtkWidget *Lable ,const char *Color,int FontSzie,const char *Word)
```
Comment (rbuj) [inline capplets/time-admin/src/time-share.c]: @raveit65 
GtkInspector: `<span foreground='gray'weight='light'font_desc='11'>Time Zone:</span>`
```shell
$ git grep SetLableFontType
src/main.c:    SetLableFontType(TimeZoneLabel,"gray",11,_("Time Zone:"));
src/main.c:    SetLableFontType(NtpSyncLabel,"gray",11,_("Ntp Sync:"));
src/main.c:    SetLableFontType(TimeLabel,"green",13,_("Set Time"));
src/main.c:    SetLableFontType(DateLabel,"green",12,_("Set Date"));
```

Comment (zhuyaliang) [reply capplets/time-admin/src/time-share.c]: The color settings have been removed
```
SetLableFontType(DateLabel,13,_("Set Date"))
```

**Comment by rbuj** [comment]
I think that maximize button should be disabled/removed from titlebar in Time And Date Manager dialog.

**Comment by raveit65** [comment]
@rbuj 
Thanks now i found it in gtk3-inspector.

**Comment by zhuyaliang** [comment]
@rbuj You're right. I'm going to remove the maximization button, too.

File: capplets/time-admin/src/time-tool.c (hunk 5 above)
Comment (rbuj) [reply capplets/time-admin/src/time-tool.c]: @zhuyaliang ping

**Comment by raveit65** [comment]
@lukefromdc 
Now we need only find out why you get the GDbus warning.

**Comment by rbuj** [comment]
@lukefromdc Try to enable/disable NTP from command line:
- Enable: `dbus-send --system --print-reply --type=method_call --dest=org.freedesktop.timedate1 /org/freedesktop/timedate1 org.freedesktop.timedate1.SetNTP boolean:true boolean:true`
- Disable: `dbus-send --system --print-reply --type=method_call --dest=org.freedesktop.timedate1 /org/freedesktop/timedate1 org.freedesktop.timedate1.SetNTP boolean:false boolean:false`

**Comment by zhuyaliang** [comment]
@lukefromdc You can also install d-feet, which allows you to graphically manipulate Dbus
2019-06-12 20-35-02屏幕截图
What about your version of system D and ntp

**Comment by lukefromdc** [comment]
I just found I did not have ntp installed at all (no surprise, my normal configuration is to block all connections to the Internet I do not deliberately make). Installed it for test, but if this doesn't do it I will step back from testing this. There is a lot of nonstandard stuff in my system, things hardcoded years ago in some cases

**Comment by lukefromdc** [comment]
Setting the clock from mate-panel's clock applet (no NTP option) works fine here. Still have the issue with or without NTP installed that the "save" button always calls up an "NTP sync" error dialog whether NTP is turned on or off. Might be my settings to use local timezone time so the BIOS time and panel time are the same.

luke@ubuntu:~$ sudo timedatectl status
               Local time: Thu 2019-06-13 01:08:02 EDT
           Universal time: Thu 2019-06-13 05:08:02 UTC
                 RTC time: Thu 2019-06-13 01:08:02
                Time zone: America/New_York (EDT, -0400)
System clock synchronized: no
              NTP service: inactive
          RTC in local TZ: yes

```
Warning: The system is configured to read the RTC time in the local time zone.
         This mode cannot be fully supported. It will create various problems
         with time zone changes and daylight saving time adjustments. The RTC
         time is never updated, it relies on external facilities to maintain it.
         If at all possible, use RTC in UTC by calling
         'timedatectl set-local-rtc 0'.
luke@ubuntu:~$ 

```

I did notice that with ntp installed, the one thing that appeared to work was turning "NTP sync" on or off, but save NEVER worked whether I ever touched that or not, and changing it had no effect on " NTP service: inactive" which may be impossible with my use of local time to keep the BIOS time and time displayed on the panel in sync. I'm not changing that, so I will probably just step back from testing this.

**Comment by lukefromdc** [comment]
To make this work no matter what, it should be able to manually set time and date when NTP is not used whether or not it is able to interact with NTP, which may or may not be installed and may not work in some configurations. When this is started with NTP turned off and the user does not attempt to turn NTP on, it should not interact at all with NTP and these errors would not then appear, at least not NTP sync errors.

**Comment by zhuyaliang** [comment]
@lukefromdc 
If you shut down the ```NTP``` service, you can change the time manually.The  ```save``` button can works. For example, you can modify your time and save it later. You can find that the local time will change.
>NTP service: inactive

NTP is off in this state. You can change your time at will. Button save can save your changes.

**Comment by zhuyaliang** [comment]
2019-06-13 15-27-49屏幕截图
2019-06-13 15-27-22屏幕截图
In this case, NTP is disabled and the time can be changed manually

**Comment by zhuyaliang** [comment]
@I found the reason why you set the time incorrectly.
First of all, my prompt misled us to ```set ntp sync```, which is incorrect. It should be ```set local time```
``` MessageReport(_("Set Local Time"),error->message,ERROR);```
He misled me into thinking that you were setting NTP sync , but in fact you were Press ```Save``` to save time，It was my fault.

The reason for the warning is the authentication problem. The GDbus interface provided by your ```systemd``` requires authentication, while other ```systemd``` does not.
You can verify this by opening the file  ```/ usr/share/polkit1/actions/org. freedesktop. timedate1. policy```
If you want to solve this problem, you need to add a Permission button.The effect is shown in the figure.What do you think of this design?
2019-06-13 18-30-55屏幕截图

**Comment by zhuyaliang** [comment]
@lukefromdc 
Can you confirm that your  ```/usr/share/polkit1/actions/org. freedesktop. timedate1. policy``` file contains ```<action id= "org. freedesktop. timedate1. set-time">```

**Comment by zhuyaliang** [comment]
@lukefromdc Now you should solve the problem of setting time. Please test it.But you need to install polkit-gobject-1

**Comment by lukefromdc** [comment]
I have and have had libpolkit-gobject-1-0 and libpolkit-gobject-1-dev (from Debian's repo installed), will try your latest change now

**Comment by lukefromdc** [comment]
@raveit65 , I accidently re-requested your review in re-requesting my own. As far as I am concerned your approval STANDS

**Comment by lukefromdc** [comment]
We now have three approvals (@raveit65's approval should be considered as still standing), but 
on Travis only the archlinux build is completing

On Debian we get 
```
No package 'polkit-gobject-1' found
at configuration
```
On Fedora we get
`Package 'polkit-gobject-1', required by 'virtual:world', not found`

On Ubuntu we get
```
configure: error: Package requirements (polkit-gobject-1) were not met:
No package 'polkit-gobject-1' found
```

Is this something we can merge now and someone familiar with Travis ensure those packages get pulled in? Many of our devs (myself included) do NOT know how to fix this kind of travis issue when a package adds a new dependency

**Comment by yetist** [comment]
@lukefromdc @zhuyaliang 

If you need to add new packages for building, you can modify the `.travis.yml` file to find the correct distro section, then add the package names and keep them in order.

For example, add a `polkit-devel` package for `fedora`:
```
diff --git a/.travis.yml b/.travis.yml
index f2de7d3..bd9aad2 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -143,6 +143,7 @@ requires:
     - marco-devel
     - mate-common
     - mate-settings-daemon-devel
+    - polkit-devel
     - redhat-rpm-config
     - startup-notification-devel
     - which
```

For debian/ubuntu, maybe this PR need `libpolkit-gobject-1-dev`.

**Comment by zhuyaliang** [comment]
@yetist  Thank you for your help. Now I've passed all the tests.

**Comment by lukefromdc** [comment]
Are we ready to merge this? It looks ready to go and this has long been missing from the control center

**Comment by zhuyaliang** [comment]
@lukefromdc I'm going to make the last change. I'm going to change the way the buttons are arranged.

**Comment by zhuyaliang** [comment]
I'm ready.

**Comment by sc0w** [comment]
This PR ads new .po files with languages, @mate-desktop/core-team this job must be done at transifex?

**Comment by raveit65** [comment]
As always, resource file for transifex needs to be updated after merging.

**Comment by raveit65** [comment]
@zhuyaliang 
Any chance to squash `delete else` into `Memory leak: g_error_free(error);` without breaking your repo ?
I hope your are familiar with the `git rebase -i commit-id-before-your-pr>` command.
If not i can do that.

**Comment by zhuyaliang** [comment]
@raveit65 I now compress two commit into one ```delete else```  into ```Memory leak: g_error_free(error);```

**Comment by raveit65** [comment]
I don`t get the logic of the lock/unlock button. Screenshot after starting the tool.
mate-time-admin
Button and icon in button says dialog is locked, but tooltip shows it is unlocked and i can change the time.
Pressing the button to unlock the tool does nothing here.
Did i over read something?

**Comment by zhuyaliang** [comment]
@raveit65  This is a normal phenomenon. The reason for this problem is that the versions of ```systemd``` are different. Some  systems authentication when they need to modify their time. At this time, the ```Lock``` works. Some ```systemd``` do not need to authenticate. At this time, the ```Lock``` does not work.

**Comment by raveit65** [comment]
Ok, good.

**Comment by sc0w** [comment]
please see warnings before merging:

clang analyzer warnings:

```
ctime-zone.c:343:5: warning: Potential leak of memory pointed to by 'db'
    g_ptr_array_foreach (db->locations, (GFunc) LoadCities, ta->CityListStore);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

time-zone.c:631:12: warning: Potential leak of memory pointed to by 'tz_info'
    offset = tz_info->utc_offset;
    ~~~~~~~^~~~~~~~~~~~~~~~~~~~~

main.c:49:21: warning: Casting a non-structure type to a structure type and accessing a field can lead to memory access errors or data corruption
    TimeAdmin *ta = (TimeAdmin *)data;
                    ^~~~~~~~~~~~~~~~~
```

cppcheck warning:

`[capplets/time-admin/src/time-tool.c:33]: (style) The scope of the variable 'LocalTime' can be reduced.`

**Comment by zhuyaliang** [comment]
I'll deal with these warnings.

**Comment by raveit65** [comment]
> please see warnings before merging:

WHy? This is your job ;)

**Comment by raveit65** [comment]
@zhuyaliang @rbuj 
How can we create a mate-time-admin.pot ?

Normally, i would add this to our ./makepot script.
```
# po-timezones for mate-time-admin
make -C capplets/time-admin/po-timezones mate-time-admin.pot &&
mv capplets/time-admin/po-timezones/mate-time-admin.pot .
```

```
./autogen.sh && ./makepot

<cut>

make: Entering directory '/media/Data-2/Programme/Linux/Mate-Desktop/git-version/github-matedesktop/mate-control-center-dist/capplets/time-admin/po-timezones'
make: *** No rule to make target 'mate-time-admin.pot'.  Stop.
make: Leaving directory '/media/Data-2/Programme/Linux/Mate-Desktop/git-version/github-matedesktop/mate-control-center-dist/capplets/time-admin/po-timezones'
```

**Comment by zhuyaliang** [comment]
@raveit65 po-timezones  The ```po-timezones```  file was copied directly from ```system-config-date```, which translated the time zone. So there is no ```pot``` creation rule.But I have the ```system-config-date-timezones.pot``` file. I don't think we need to generate pot files dynamically from ```Makefile```. The content of this file will not change. We can put ```mate-c-c-timezones.pot```  manually.
[mate-control-center-timezones.pot.tar.gz](https://github.com/mate-desktop/mate-control-center/files/3293380/mate-control-center-timezones.pot.tar.gz)

**Comment by raveit65** [comment]
True, i think we can expect that nobody will remove or add a new timezone :)
I will include the (static) .pot file.

**Comment by rbuj** [comment]
@raveit65 Time Zones are maintained in https://www.iana.org/time-zones
L10N for time zones: https://github.com/mate-desktop/mate-control-center/pull/462
```shell
$ git grep TZ_DATA_FILE
time-zone.c:    file = g_strdup (TZ_DATA_FILE);
time-zone.h:#  define TZ_DATA_FILE "/usr/share/zoneinfo/zone.tab"
time-zone.h:#  define TZ_DATA_FILE "/usr/share/lib/zoneinfo/tab/zone_sun.tab"
$ rpm -qf /usr/share/zoneinfo/zone.tab
tzdata-2019a-1.fc30.noarch
$ LANG=C rpm -qi tzdata-2019a
Name        : tzdata
Version     : 2019a
Release     : 1.fc30
Architecture: noarch
Install Date: Wed Apr 17 16:46:51 2019
Group       : Unspecified
Size        : 1882000
License     : Public Domain
Signature   : RSA/SHA256, Sat Mar 30 03:58:53 2019, Key ID ef3c111fcfc659b9
Source RPM  : tzdata-2019a-1.fc30.src.rpm
Build Date  : Sat Mar 30 03:37:48 2019
Build Host  : buildvm-ppc64le-10.ppc.fedoraproject.org
Relocations : (not relocatable)
Packager    : Fedora Project
Vendor      : Fedora Project
URL         : https://www.iana.org/time-zones
Bug URL     : https://bugz.fedoraproject.org/tzdata
Summary     : Timezone data
Description :
This package contains data files with rules for various timezones around
the world.

```

**Comment by raveit65** [comment]
https://github.com/mate-desktop/mate-control-center/pull/462

**Pull request merged**